src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,642 @@
+            set.index.unset(index_bit);
+        }
+
         const address = bit + 1;
         return @intCast(u64, address);
     }
 
+    fn try_acquire_recent(set: *BlockFreeSet) ?u64 {
+        const bit = set.recent.pop() orelse return null;
+        assert(set.blocks.isSet(bit));
+        set.blocks.unset(bit);
+        const shard = @divTrunc(bit, shard_size);
+        const shard_start = shard * shard_size;
+        const shard_end = std.math.min(shard_start + shard_size, set.blocks.bit_length);
+        if (findFirstSetBit(set.blocks, shard_start, shard_end) == null) set.index.unset(shard);
+        return @intCast(u64, bit + 1);
+    }
+
+    fn isFree(set: *BlockFreeSet, address: u64) bool {
+        return set.blocks.isSet(address - 1);
+    }
+
+    // Mark the specified block as free.
     pub fn release(set: *BlockFreeSet, address: u64) void {
         const bit = address - 1;
-        assert(!set.free.isSet(bit));
-        set.free.set(bit);
+        assert(!set.blocks.isSet(bit));
+
+        set.blocks.set(bit);
+        set.index.set(@divTrunc(bit, shard_size));
+        set.recent.push(bit) catch {};
+    }
+
+    pub fn decode(data: []const u8, allocator: *mem.Allocator) !BlockFreeSet {
+        var blocks = try BitsetEncoder(.immutable).decode(data, allocator);
+        errdefer blocks.deinit(allocator);
+
+        const index_size = divCeil(usize, blocks.bit_length, shard_size);
+        var index = try DynamicBitSetUnmanaged.initFull(index_size, allocator);
+        errdefer index.deinit(allocator);
+
+        var i: usize = 0;
+        while (i < index_size) : (i += 1) {
+            const shard_start = i * shard_size;
+            const shard_end = std.math.min(shard_start + shard_size, blocks.bit_length);
+            if (findFirstSetBit(blocks, shard_start, shard_end) == null) index.unset(i);
+        }
+
+        return BlockFreeSet{
+            .index = index,
+            .blocks = blocks,
+        };
+    }
+
+    // Returns the number of bytes that the `BlockFreeSet` needs to encode to.
+    pub fn encodeSize(set: BlockFreeSet) usize {
+        return set.makeEncoder(.immutable).offsets.end;
+    }
+
+    pub fn encode(set: BlockFreeSet, dst: []u8) void {
+        set.makeEncoder(.mutable).encode(dst, set.blocks);
+    }
+
+    fn makeEncoder(set: BlockFreeSet, comptime M: Mutable) BitsetEncoder(M) {
+        const words = bitsetMasksLen(set.blocks);
+        var runs: usize = 0;
+        var literals: usize = 0;
+        var i: usize = 0;
+        while (i < words) {
+            const word = set.blocks.masks[i];
+            if (word == 0 or word == ~@as(usize, 0)) {
+                runs += 1;
+                i += 1 + bitsetRunLength(set.blocks, i + 1, words, word);
+            } else {
+                literals += 1;
+                i += 1;
+            }
+        }
+        return BitsetEncoder(M).init(.{
+            .runs = runs,
+            .literals = literals,
+        });
```

Comment:
`makeEncoder` is also currently used when predicting the encoded buffer size, so the caller can preallocate the right amount of space. What pattern are you imagining that the caller will use instead? e.g. they could allocate the upper-bound size (as if there were no runs, only literals) and then `encode` would return the number if bytes it actually used?

Reply:
Yes, exactly. Because we statically allocate all buffers that TB will ever use during runtime, upfront when starting, and because we never allocate anything once runtime has then started, we would always be provided a max-sized buffer here when encoding. We would basically have a block_free_set_buffer_max or something like that (i.e. max addressable local blocks total for any TB cluster, which would be something fairly large).

Reply:
Great, that should simplify the encoding logic as well :tada: